Increase Compatibility of NETBALAN Restart Output

bska opened the pull request:
The current `Network::Balance` class applies default values a little too eagerly and this leads to problems for restarted simulation.  We must take care to preserve the "did this item have a value" state also in the restart file lest interactions with `TUNING` be lost there.

To this end, make the minimum timestep size for network balancing be an `std::optional<>` and use `std::nullopt` to represent "no value set" in the input file.  While here, also preserve negative balancing intervals exactly instead of replacing these values by zero.

bska commented:
@tskille: I believe this is ready for testing on your field case now.  For testing convenience this PR currently also includes #3021 and must therefore not be merged until that PR is installed on the master branch.  I am keeping this PR in "draft" state until that happens.

tskille commented:
This had now been tested on a full field model including advanced features including network option. 4 different restart steps has been tested and results are excellent.

bska commented:
Thank you, @tskille, for testing the PR.  With those testing results in hand I'm marking this as "ready for review" and running the first build test.  I expect a couple of regression failures and I will verify that those are only of the expected form.

bska commented:
jenkins build this please

bska commented:
jenkins build this please

bska commented:
jenkins build this please

blattms reviewed (approved):
assuming that the algorithm changes are correct, this looks good. Just one minor nit-pick (hard to find when enum classes are zero-initialized)

bska commented:
> Just one minor nit-pick (hard to find when enum classes are zero-initialized)

Thanks for the review.  I've pushed an update which addresses your comments to this PR now.

bska commented:
jenkins build this please
